### DISCUSSION TITLE
How to perform maintenance on Javers mysql tables?

### DISCUSSION THREAD

**Discussion by aquacode** [javers/javers#1352] [Q&A] [answered]
Documentation and online help is sparse on the topic of purging, cleaning, deleting or archiving Javers data in Mysql.
I see a few posts stating delete from jv_snapshot where etc but it doesn't have a comprehensive explanation of how the tables are to be purged of all data from a certain date backwards. 
The reason we want to do this is because some of our environments are of course for dev and testing. We don't want this data to grow indefinitely.
Is there any canonical documentation on how to manage this that someone can point me to?

**Comment by bartoszwalacik** [comment, marked as answer]
There is no doc for this subject but Javers' tables are normal tables with straightforward schema, see [FixedSchemaFactory.java](https://github.com/javers/javers/blob/master/javers-persistence-sql/src/main/java/org/javers/repository/sql/schema/FixedSchemaFactory.java )

There are four tables:
- jv_snapshot
- jv_commit
- jv_commit_property
- jv_global_id

Procedure:
1. identify a certain date (X), all commits before that date will to be purged 
1. `delete from jv_commit_property where commit_fk in (select commit_pk from jv_commit where commit_date < X)`
1. `delete from jv_snapshot where commit_fk in (select commit_pk from jv_commit where commit_date < X)`
1. `delete from jv_commit where commit_date < X`
1. `delete from jv_global_id where not exists (select 1 from jv_snapshot where global_id_fk = global_id_pk)`

**Reply by aquacode** [reply]
I finally got around to testing this since before you answered I had gone off on my own and figured out how to clean up the tables. I ended up with something very similar to your answer except for the jv_global_id table. I was getting foreign key constraint errors.

In fact, I'm getting them using your command on line 5 above, and also using my own command:
`DELETE FROM jv_global_id
WHERE NOT EXISTS (SELECT * FROM jv_snapshot WHERE jv_snapshot.global_id_fk = jv_global_id.global_id_pk);`

Using your command, or mine, I'm getting the following output:

`ERROR 1451 (23000): Cannot delete or update a parent row: a foreign key constraint fails (`db`.`jv_global_id`, CONSTRAINT `jv_global_id_owner_id_fk` FOREIGN KEY (`owner_id_fk`) REFERENCES `jv_global_id` (`global_id_pk`))`

Neither seems to be working now, any ideas?

**Reply by aquacode** [reply]
Apparently this works (two separate delete statements):

`delete from jv_global_id where owner_id_fk is not null and not exists (select 1 from jv_snapshot where global_id_fk = global_id_pk);`
`delete from jv_global_id where not exists (select 1 from jv_snapshot where global_id_fk = global_id_pk);`

First takes care of the child rows and then cleans up the parents.

**Comment by aquacode** [comment]
Thanks, very helpful!
